server/src/main/java/org/opensearch/common/io/VersionedCodecStreamWrapper.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.io;
+
+import java.io.IOException;
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.store.BufferedChecksumIndexInput;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+
+/**
+ * Manages versioning and checksum for a stream of content.
+ * @param <T> Type of content to be read/written
+ *
+ * @opensearch.internal
+ */
+public class VersionedCodecStreamWrapper<T> {
```

Comment:
I still do not find naming to be intuitive but don't have better suggestions.

Reply:
We need tests for this newly created classes

Reply:
The code flow is all covered in RemoteSegmenStoreDirectoryTests.java. 

But since VersionedCodecStreamWrapper class is generic, we can added UTs for this. 

Other new classes are segment store related so we should be good with coverage in RemoteSegmentStoreDirectoryTests? or do you feel we need UTs for those cases as well?

Reply:
Added UTs for both `VersionedCodecStreamWrapper` and `RemoteSegmentMetadataHandler`.